### DISCUSSION TITLE
Some questions about coil coordinate system definition

### DISCUSSION THREAD

**Discussion by windyjunfeng** [simnibs/simnibs#500] [8. Simulation workflows]
SimNIBS is a great software, but I have some questions about the definition of coil coordinate system according to the center of coil and the yaw angle of coil.
I find the corresponding codes in simnibs/optimization/optimize_tms.py.
**get_opt_grid function**
```       
        z = -n
        y = y_seed - (z * y_seed.dot(z))
        y /= np.linalg.norm(y)
        x = np.cross(y, z)
        R = np.array([x, y, z]).T
        rotated = _rotate_system(R, angle_limits, resolution_angle)
```

1. According to the codes, if the yaw angle is 0, R will be the rotation matrix of the coil. If the handle reference direction is not specified, y_seed will be np.array([0., 1., 0.]). When the yaw angle is 0 and the handle reference direction is not specified (that is the default value np.array([0., 1., 0.])), I visualize the y axis of the coil coordinate system.
![image](https://github.com/simnibs/simnibs/assets/58356387/69a5b537-ab93-4dea-bfa6-e833e8940bfd)
It can be seen that the y axis of the coil (the blue line) doesn't locate in the Y-Z plane in world coordinates in subject space. However, if the yaw angle is 0, the y axis of the coil  should generally locate in the Y-Z plane in world coordinates in subject space or in the median sagittal plane. So what's the actual meaning of the reference direction? Could you please give some reasonable explanation? And why is the default value of y_seed np.array([0., 1., 0.])?
2. If the z axis of coil coordinate system is parallel to the y axis in world coordinates in subject space (that is z=(0,0,1) or z=(0,0,-1)), y wil be (0,0,0). In this case, y /= np.linalg.norm(y) will occur the error because the denominator np.linalg.norm(y) is zero. So, I think some extreme situations should be considered.
3. According to the codes, the positive z axis of coil coordinate system is perpendicular to the scalp inward and the positive y axis of coil coordinate system is the handle of the coil. However, the positive z-axis and the positive y-axis in the coil file in your website are the opposite of what's in codes. The file may need correction.
![image](https://github.com/simnibs/simnibs/assets/58356387/321c7d6f-361b-48aa-96ba-93e3f53bcf0e)

I think this section can be implemented as follows.
```
z = -n
if z[2]==0:
     y=np.array([0,0,-1])
else:
     y=np.array([0,-1,z[1]/z[2]])
y /= np.linalg.norm(y)
x = np.cross(y, z)
R = np.array([x, y, z]).T
rotated = _rotate_system(R, angle_limits, resolution_angle)
```
The main thought is as follows.
If the yaw angle θ is 0, the projection of the positive y axis of coil coordinate system onto the x axis in world coordinates in subject space is np.sin(θ)=0. And the projection of the positive y axis of coil coordinate system onto the y axis in world coordinates in subject space is -np.cos(θ)=-1. (Assume that the positive directions of axes are consistent with the original codes, not with the file.) Since the y-axis is perpendicular to the z-axis, the dot product of y and z is 0. So, it can be calculated that the projection of the positive y axis of coil coordinate system onto the z axis in world coordinates in subject space is z[1]/z[2] (z[2]=0 is the extreme situation). R is the rotation matrix when yaw angle is 0.

Sincerely look forward to your reply! Thank you~

**Comment by oulap** [comment]
The y-direction of the coil is defined relative to the center of the coil, i.e., in the default case the y-direction will be [0, 1, 0], which points away from the handle. If you want to define a direction for the coil, you should pick a point somewhere on the scalp, and the direction is then defined as `reference_point - coil_center`. In case you want to just define a direction, you can pass the reference point as: `coil_center + direction`. 

The way the coil coordinate system is spanned is a in the code you pasted:
```
        z = -n
        y = y_seed - (z * y_seed.dot(z))
        y /= np.linalg.norm(y)
        x = np.cross(y, z)
        R = np.array([x, y, z]).T
        rotated = _rotate_system(R, angle_limits, resolution_angle)
```

First the normal component of the skin is extracted (n) and inverted (-n), so the active side of the coil point towards the head. Then the "y_seed" which is the direction relative to the coil center is projected to be orthogonal to z and normalized. Finally the x-direction is defined normal to both y and z, which defines the coordinate system. You can also look at this function, which is documented slightly better:

https://github.com/simnibs/simnibs/blob/e15afab5712838fe728e62f29505172f8d096fd8/simnibs/mesh_tools/mesh_io.py#L1877

**Comment by windyjunfeng** [comment]
@oulap Thank you very much. I understand what you said and read the ```calc_matsimnibs function```, but I still have confusion about the real meaning of "the direction relative to the coil center". 

The coil has 6 degrees of freedom, i.e., x, y, z, the pitch angle, the roll angle, the yaw angle. Due to coil is tangent to the scalp, the angle in the code can be regarded as the yaw angle in the original coil coordinate system (when the angle is set to 0). Generally speaking, when the angle is set 0, the coil handle is parallel to Y-Z plane in the subject space. However, in the codes, the coil handle may not be parallel to Y-Z plane, because the x-axis component of ```z * y_seed.dot(z)``` may not be 0 when y_seed is [0,1,0]. So, I think it's very hard to place the coil at the optimal angle computed by SimNIBS without the navigation system, because the reference direction (that is when angle is set to 0) cannot be located intuitively by human eyes. In other words, y (not y_seed) actually represents the reference direction, that is the handle direction when angle is set to 0.

Besides, I think when ```z * y_seed.dot(z)``` is equal to y_seed, y will be 0. In this extreme case, ```y /= np.linalg.norm(y)``` will occur error. For example, if z is [0,1,0] according to the computation and y_seed is [0,1,0], the error will occur. (I'm sorry that there is a mistake in my original description.)

I think my codes in the original description can handle the problems. When the angle is 0, the coil handle is parallel to Y-Z plane in the subject space in any cases. The following picture is the y direction (coil handle direction) according to my codes when angle is set to 0.
![微信图片_20240129184556](https://github.com/simnibs/simnibs/assets/58356387/80849183-bde6-48c8-b252-bda3fd2e0c6b)

Maybe I still misunderstand what you said. Sincerely look forward to your reply~
